Add compiler configuration

michaeljones opened the pull request:
With an error-format specification which attempts to extract the appropriate information out of a standard gleam error message.

This the standard way to populate the quickfix list to make it easer to navigate between them with standard shortcuts.
    
Initially based on the rust vim compiler file:
    
      https://github.com/rust-lang/rust.vim/blob/master/compiler/rustc.vim
    
Though I'm not sure how much we need to bring across.

michaeljones committed:
Add compiler configuration

With an error-format specification which attempts to extract the
appropriate information out of a standard gleam error message.

Initially based on the rust vim compiler file:

  https://github.com/rust-lang/rust.vim/blob/master/compiler/rustc.vim

Though I'm not sure how much we need to bring across.

lpil commented:
What a cool idea! Could you share how this is used?

The CLI output is not a reliable API as it will change. We could add a fixed format for this, one that is easier to parse.

michaeljones commented:
So it ties into the vim `:make` functionality. I've got a personal `.vim/ftplugin/gleam.vim` file with:

```
" Gleam Specific
" --------------

" Only do this when not done yet for this buffer
if exists("b:did_ftplugin")
  finish
endif

compiler gleam
map <leader>m :make build .<CR>
```
This tells vim to use the `gleam` compiler configuration which I guess gets it to go and source the `compiler/gleam.vim` file in this PR. This file sets it so that when I run `:make build .` it runs `gleam build .` and then it uses the error format we've specified by setting `efm` to parse the error messages from gleam and populate the quickfix list with files, lines, columns and error messages and allows you to jump to the first error with `:cc` and to the last/next one with `:cp`/`:cn`. It actually jumps to the first error automatically which is nice.

Writing this I imagine that it would be a little clearer to use 'errorformat` instead of the short hand `efm`. it is possible that it would also be better with `setlocal` instead of `set`. I am not a vim expert :/

The errorformat string is like a little (horrible) parsing language and so, yeah, it'll have to be changed if the error format changes. I've also not tested the functionality on a situation with more than one error. I'm very new to gleam and couldn't find a combination of mistakes that would create more than one error. That said, writing this, I realise I could have made mistakes in multiple different files and seen how that was handled.

lpil commented:
> I'm very new to gleam and couldn't find a combination of mistakes that would create more than one error.

At present we only return a single error.

Thank you for the detail! Would you mind adding some comments to the file to that effect as I'm likely going to forget and get confused later 😅

michaeljones commented:
Understandable. It is certainly a dark corner of things. Best to document it as well as possible.

michaeljones committed:
Add explanatory comments & use CompilerSet

I'm not sure of when to use let vs set vs setlocal but most compiler
files seem to define (if needed) and then use CompilerSet this way.

Setting the rules as strings is easier to read as you don't have to
escape the spaces but I'm not sure how to get it working with
CompilerSet so we just go with it. It is probably quite basic but I get
confused.

lpil reviewed (approved):
Thank you ✨

michaeljones commented:
No trouble. Thanks for being so quick :)
